cmd/util/ledger/migrations/core_contracts_migration.go

```diff
@@ -0,0 +1,265 @@
+			// flowTokenAddress:
+			addresses.flowToken,
+		),
+	)
+	addCode(
+		addresses.qc,
+		"FlowClusterQC",
+		coreContracts.FlowQC(),
+	)
+	addCode(
+		addresses.dkg,
+		"FlowDKG",
+		coreContracts.FlowDKG(),
+	)
+	addCode(
+		addresses.epoch,
+		"FlowEpoch",
+		coreContracts.FlowEpoch(
+			// fungibleTokenAddress:
+			addresses.fungibleToken,
+			// flowTokenAddress:
+			addresses.flowToken,
+			// idTableAddress:
+			addresses.stakingTable,
+			// qcAddress:
+			addresses.qc,
+			// dkgAddress:
+			addresses.dkg,
+			// flowFeesAddress:
+			addresses.fees,
+		),
+	)
+	addCode(
+		addresses.stakingTable,
+		"FlowIDTableStaking",
+		coreContracts.FlowIDTableStaking(
+			// fungibleTokenAddress:
+			addresses.fungibleToken,
+			// flowTokenAddress:
+			addresses.flowToken,
+			// flowFeesAddress:
+			addresses.fees,
+			// latest:
+			true,
+		),
+	)
+	addCode(
+		addresses.lockedTokens,
+		"LockedTokens",
+		coreContracts.FlowLockedTokens(
+			// fungibleTokenAddress:
+			addresses.fungibleToken,
+			// flowTokenAddress:
+			addresses.flowToken,
+			// idTableAddress:
+			addresses.stakingTable,
+			// stakingProxyAddress:
+			addresses.stakingProxy,
+			// storageFeesAddress:
+			addresses.storageFees,
+		),
+	)
+	addCode(
+		addresses.stakingCollection,
+		"FlowStakingCollection",
+		coreContracts.FlowStakingCollection(
+			// fungibleTokenAddress:
+			addresses.fungibleToken,
+			// flowTokenAddress:
+			addresses.flowToken,
+			// idTableAddress:
+			addresses.stakingTable,
+			// stakingProxyAddress:
+			addresses.stakingProxy,
+			// lockedTokensAddress:
+			addresses.lockedTokens,
+			// storageFeesAddress:
+			addresses.storageFees,
+			// qcAddress:
+			addresses.qc,
+			// dkgAddress:
+			addresses.dkg,
+			// epochAddress:
+			addresses.epoch,
+		),
+	)
+	addCode(
+		addresses.flowToken,
+		"FlowToken",
+		coreContracts.FlowToken(
+			// fungibleTokenAddress:
+			addresses.fungibleToken,
+		),
+	)
+	addCode(
+		addresses.stakingProxy,
+		"StakingProxy",
+		coreContracts.FlowStakingProxy(),
+	)
+
+	return codes
+}
+
+const codeKeyPrefix = "code."
+
+func (m *CoreContractsMigration) Migrate(payloads []ledger.Payload) ([]ledger.Payload, error) {
+
+	m.Updates = map[ContractLocation]struct{}{}
+
+	m.Log.Info().Msgf("Running Core Contracts migration")
+
+	addresses := coreContractAddresses(m.Chain)
+	codes := coreContractCodes(addresses)
+
+	for i, payload := range payloads {
+		key := string(payload.Key.KeyParts[2].Value)
+		if !strings.HasPrefix(key, codeKeyPrefix) {
+			continue
+		}
+
+		owner := hex.EncodeToString(payload.Key.KeyParts[0].Value)
+
+		addressCodes, ok := codes[owner]
+		if !ok {
+			continue
+		}
+
+		contractName := strings.TrimPrefix(key, codeKeyPrefix)
+
+		newCode, ok := addressCodes[contractName]
+		if !ok {
+			m.Log.Warn().Msgf("unknown contract: %s", contractName)
+			continue
+		}
+
+		payloads[i].Value = newCode
+
+		m.Log.Info().Msgf("Updated core contract: %s.%s", owner, contractName)
+
```

Comment:
do we want to log the diff between old code and new code to a file or log in case we have a unexpected change?

Reply:
Good idea, I'll add that!

Reply:
Added in [dca8edb](https://github.com/onflow/flow-go/pull/2578/commits/dca8edbe3e192e5a30509ba76a190da8e1cde578)